x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/ApproximateRegExp.java

```diff
@@ -0,0 +1,631 @@
+        to = e.to;
+    }
+
+    // Convert a regular expression to a simplified query consisting of BooleanQuery and TermQuery objects
+    // which captures as much of the logic as possible. Query can produce some false positives but shouldn't
+    // produce any false negatives.
+    // In addition to Term and BooleanQuery clauses there are MatchAllDocsQuery objects (e.g for .*) and
+    // an equivalent MatchAllButRequireVerificationQuery. 
+    // *  If an expression resolves to a single MatchAllDocsQuery eg .* then a match all shortcut is possible with 
+    //    no verification needed.     
+    // * If an expression resolves to a MatchAllButRequireVerificationQuery eg ?? then only the verification 
+    //   query is run.
+    // * Anything else is a concrete query that should be run on the ngram index.
+    public Query toApproximationQuery(StringNormalizer normalizer) throws IllegalArgumentException {
+        Query result = null;
+        switch (kind) {
+            case REGEXP_UNION:
+                result = createUnionQuery(normalizer);
+                break;
+            case REGEXP_CONCATENATION:
+                result = createConcatenationQuery(normalizer);
+                break;
+            case REGEXP_STRING:
+                String normalizedString = normalizer == null ? s : normalizer.normalize(s);
+                result = new TermQuery(new Term("", normalizedString));
+                break;
+            case REGEXP_CHAR:
+                String cs =Character.toString(c);
+                String normalizedChar = normalizer == null ? cs : normalizer.normalize(cs);
+                result = new TermQuery(new Term("", normalizedChar));
+                break;
+            case REGEXP_REPEAT:
+                // Repeat is zero or more times so zero matches = match all
+                result = new MatchAllDocsQuery();
+                break;
+                
+            case REGEXP_REPEAT_MIN:
+            case REGEXP_REPEAT_MINMAX:
+                if (min > 0) {
+                    result = exp1.toApproximationQuery(normalizer);
+                    if(result instanceof TermQuery) {
+                        // Wrap the repeating expression so that it is not concatenated by a parent which concatenates
+                        // plain TermQuery objects together. Boolean queries are interpreted as a black box and not
+                        // concatenated.
+                        BooleanQuery.Builder wrapper = new BooleanQuery.Builder();
+                        wrapper.add(result, Occur.MUST);
+                        result = wrapper.build();
+                    }
+                } else {
+                    // TODO match all was a nice assumption here for optimising .* but breaks 
+                    // for (a){0,3} which isn't a logical match all but empty string or up to 3 a's. 
+//                    result = new MatchAllDocsQuery();
+                    result = new MatchAllButRequireVerificationQuery();
```

Comment:
I thought this class made the intention much clearer. Without it we'd have to interpret any FieldExistsQuery to mean "don't run a verification" query  (e.g. for regex `.*`) and for a MatchAllDocsQuery to mean DO run a verification query (e.g. for regex `..`).
There are certain queries like `.*` where we know we can satisfy all criteria using the ngram index only.  However I thought today that we could extend this support by making MatchAllButRequireVerificationQuery a base class or marker interface. The regex `a` for example could also be satisfied using the ngram index only so returning a form of term query which tested for true in `instanceof MatchAllButRequireVerificationQuery` would help the WildcardField avoid running pointless verification for more than one regex type.

Reply:
I see but is it still the case if the ngram index is lowercased ? The search is case-sensitive so unless we differentiate characters eligible for case folding, we have to check every match ?

Reply:
We can see if the search value changes after normalisation and only optimise the query if there's no change.

Reply:
> We can see if the search value changes after normalisation and only optimise the query if there's no change.
Perhaps a rare example of a user regex but illustrates the need to communicate where we think we can handle queries with the ngram index only.

I don't think this works since values in documents would be lowercased too so you'd need a verification for any character that can upper and lower cased ? We can think about optimizations in the future but as you said in previous comments we should aim for correctness first.